apps/api/src/app/inbox/usecases/snooze-notification/snooze-notification.usecase.ts

```diff
@@ -0,0 +1,243 @@
+  CreateExecutionDetailsCommand,
+  DetailEnum,
+  StandardQueueService,
+  FeatureFlagsService,
+  SYSTEM_LIMITS,
+  PinoLogger,
+} from '@novu/application-generic';
+import {
+  JobEntity,
+  JobRepository,
+  MessageRepository,
+  MessageEntity,
+  OrganizationEntity,
+  EnvironmentEntity,
+  UserEntity,
+  CommunityOrganizationRepository,
+} from '@novu/dal';
+import {
+  ApiServiceLevelEnum,
+  ChannelTypeEnum,
+  ExecutionDetailsSourceEnum,
+  ExecutionDetailsStatusEnum,
+  FeatureFlagsKeysEnum,
+  FeatureNameEnum,
+  getFeatureForTierAsNumber,
+  JobStatusEnum,
+} from '@novu/shared';
+import { Types } from 'mongoose';
+import { v4 as uuidv4 } from 'uuid';
+import { SnoozeNotificationCommand } from './snooze-notification.command';
+import { MarkNotificationAs } from '../mark-notification-as/mark-notification-as.usecase';
+import { MarkNotificationAsCommand } from '../mark-notification-as/mark-notification-as.command';
+import { InboxNotification } from '../../utils/types';
+
+const LOG_CONTEXT = 'SnoozeNotification';
+
+@Injectable()
+export class SnoozeNotification {
+  private readonly RETRY_ATTEMPTS = 3;
+
+  constructor(
+    private readonly logger: PinoLogger,
+    private messageRepository: MessageRepository,
+    private jobRepository: JobRepository,
+    private standardQueueService: StandardQueueService,
+    private organizationRepository: CommunityOrganizationRepository,
+    private createExecutionDetails: CreateExecutionDetails,
+    private markNotificationAs: MarkNotificationAs,
+    private featureFlagsService: FeatureFlagsService
+  ) {}
+
+  public async execute(command: SnoozeNotificationCommand): Promise<InboxNotification> {
+    await this.isSnoozeEnabled(command);
+
+    const notification = await this.findNotification(command);
+    const delayAmount = this.calculateDelayInMs(command.snoozeUntil);
+    await this.validateDelayDuration(command, delayAmount);
+
+    try {
+      let scheduledJob = {} as JobEntity;
+      let snoozedNotification = {} as InboxNotification;
+
+      await this.messageRepository.withTransaction(async () => {
+        scheduledJob = await this.createScheduledUnsnoozeJob(notification, delayAmount);
+        snoozedNotification = await this.markNotificationAsSnoozed(command);
+        await this.queueJob(scheduledJob, delayAmount);
+      });
+
+      // fire and forget
+      this.createExecutionDetails
+        .execute(
+          CreateExecutionDetailsCommand.create({
+            ...CreateExecutionDetailsCommand.getDetailsFromJob(scheduledJob),
+            detail: DetailEnum.MESSAGE_SNOOZED,
+            source: ExecutionDetailsSourceEnum.INTERNAL,
+            status: ExecutionDetailsStatusEnum.PENDING,
+            isTest: false,
+            isRetry: false,
+          })
+        )
+        .catch((error) => {
+          this.logger.error({ err: error }, 'Failed to create execution details', LOG_CONTEXT);
+        });
+
+      return snoozedNotification;
+    } catch (error) {
```

Comment:
I don't think we need try/catch here. If there is an error, it should bubble up.

Reply:
The idea here is that if the error happens outside of try/catch, that means its a user input/validation/tier error which doesn't need an action.

If it happens in the specific try/catch part then we know its our fault = internal error and we need to see it in logs clearly, otherwise e.g. `mark-notification-as` usecase (which is in the try/catch) can throw `BadRequestError` which I would likely ignore when seen in NewRelic, but now it would be transformer to `InternalErrorException`.